src/block.rs

```diff
@@ -1,101 +1,428 @@
+/// at most once. All subsequent calls will use a cached version.
+pub struct BlockGeneric<C, H>
 where
-    C: Into<u64> + TryFrom<u64> + Copy,
-    H: Into<u64> + TryFrom<u64> + Copy,
+    C: Copy + TryFrom<u64> + Into<u64> + EncodeDecodeIpld<H>,
+    H: Copy + TryFrom<u64> + Into<u64>,
 {
-    /// Content identifier.
-    pub cid: CidGeneric<C, H>,
-    /// Binary data.
-    pub data: Box<[u8]>,
+    cid: Option<CidGeneric<C, H>>,
+    raw: Option<Box<[u8]>>,
+    node: Option<Ipld<C, H>>,
+    codec: C,
+    hash_alg: H,
 }
 
-/// Encode a block.`
-pub fn encode<C, H, O, M, E>(e: &E) -> Result<Block<C, H>>
+impl<C, H> BlockGeneric<C, H>
+where
+    C: Copy + TryFrom<u64> + Into<u64> + EncodeDecodeIpld<H>,
+    H: Copy + TryFrom<u64> + Into<u64>,
+{
+    /// Create a new `Block` from the given CID and raw binary data.
+    ///
+    /// It needs a registry that contains codec and hash algorithms implementations in order to
+    /// be able to decode the data into IPLD.
+    pub fn new(cid: CidGeneric<C, H>, raw: Box<[u8]>) -> Self {
+        let codec = cid.codec();
+        let hash_alg = cid.hash().algorithm();
+        Self {
+            cid: Some(cid),
+            raw: Some(raw),
+            node: None,
+            codec,
+            hash_alg,
+        }
+    }
+
+    /// Create a new `Block` from the given IPLD object, codec and hash algorithm.
+    ///
+    /// No computation is done, the CID creation and the encoding will only be performed when the
+    /// corresponding methods are called.
+    pub fn encoder(node: Ipld<C, H>, codec: C, hash_alg: H) -> Self {
+        Self {
+            cid: None,
+            raw: None,
+            node: Some(node),
+            codec,
+            hash_alg,
+        }
+    }
+
+    /// Create a new `Block` from encoded data, codec and hash algorithm.
+    ///
+    /// No computation is done, the CID creation and the decoding will only be performed when the
+    /// corresponding methods are called.
+    pub fn decoder(raw: Box<[u8]>, codec: C, hash_alg: H) -> Self {
+        Self {
+            cid: None,
+            raw: Some(raw),
+            node: None,
+            codec,
+            hash_alg,
+        }
+    }
+
+    /// Decode the `Block` into an IPLD object.
+    ///
+    /// The actual decoding is only performed if the object doesn't have a copy of the IPLD
+    /// object yet. If that method was called before, it returns the cached result.
+    pub fn decode(&mut self) -> Result<&Ipld<C, H>, BlockError> {
+        if let Some(ref node) = self.node {
+            Ok(node)
+        } else if let Some(ref raw) = self.raw {
+            let decoded = self
+                .codec
+                .decode(raw)
+                .map_err(|err| BlockError::DecodeError(err.to_string()))?;
+            self.node = Some(decoded);
+            Ok(self.node.as_ref().unwrap())
+        } else {
+            Err(BlockError::DecodeError(
```

Comment:
These are irrecoverable programmer errors. I think an immutable api could be built to make sure these can't happen at compile time.